File: gfapy/line/group/path/references.py
import gfapy

class References:

  def _compute_required_links(self):
    """
    Computes the list of links which are required to support
    the path.

    Returns
    -------
    list of (gfapy.OrientedLine, gfapy.OrientedLine, gfapy.CIGAR)
      An array, which elements are 3-tuples
      (from oriented segment, to oriented segment, cigar)
    """
    if len(self.segment_names) == 1:
      return []
    has_undef_overlaps = self._undef_overlaps()
    retval = []
    is_circular = self.is_circular()
    for i in range(len(self.segment_names)):
      j = i+1
      if j == len(self.segment_names):
        if is_circular:
          j = 0
        else:
          break
      if has_undef_overlaps:
        cigar = gfapy.AlignmentPlaceholder()
      else:
        cigar = self.overlaps[i]
      retval.append([self.segment_names[i], self.segment_names[j], cigar])
    return retval

  def _undef_overlaps(self):
    """
    Are the overlaps a single "*"? This is a compact representation of
    a linear path where all CIGARs are "*".

    Returns
    -------
    bool
    """
    return len(self.overlaps) == 1 and gfapy.is_placeholder(self.overlaps[0])

  def _initialize_references(self):
    self._initialize_links()
    self._initialize_segments()

  def _initialize_links(self):
    self._refs["links"] = []
    for from_segment, to_segment, cigar in self._compute_required_links():
      l = None
      orient = "+"
      if self._gfa.segment(from_segment.line) and self._gfa.segment(to_segment.line):
        l = self._gfa._search_link(from_segment, to_segment, cigar)
        if l is not None and l.is_compatible_complement(from_segment, to_segment, cigar):
          orient = "-"
      if l is None:
        if self._gfa._segments_first_order:
          raise gfapy.NotFoundError("Path: {}\n".format(self)+
          "requires a non-existing link:\n"+
          "from={} to={} cigar={}".format(from_segment, to_segment, cigar))
        l = gfapy.line.edge.Link({"from_segment" : from_segment.line,
                                  "from_orient" : from_segment.orient,
                                  "to_segment" : to_segment.line,
                                  "to_orient" : to_segment.orient,
                                  "overlap" : cigar},
                                 virtual = True,
                                 version = "gfa1")
        l.connect(self._gfa)
      self._refs["links"].append(gfapy.OrientedLine(l,orient))
      l._add_reference(self, "paths")

  def _initialize_segments(self):
    for sn_with_o in self.segment_names:
      s = self._gfa.segment(sn_with_o.line)
      sn_with_o.line = s
      s._add_reference(self, "paths")

  def _backreference_keys(self, ref, key_in_ref):
    if ref.record_type == "L":
      return ["links"]
    elif ref.record_type == "S":
      return ["segment_names"]

File: lib/rgfa/line/group/path/references.rb
module RGFA::Line::Group::Path::References

  private

  # Computes the list of links which are required to support
  # the path
  # @return
  #   [Array<[RGFA::OrientedLine, RGFA::OrientedLine, RGFA::Alignment::CIGAR]>]
  #   an array, which elements are 3-tuples (from oriented segment,
  #   to oriented segment, cigar)
  # @api private
  def compute_required_links
    has_undef_overlaps = undef_overlaps?
    retval = []
    is_circular = self.circular?
    segment_names.size.times do |i|
      j = i+1
      if j == self.segment_names.size
        is_circular ? j = 0 : break
      end
      cigar = has_undef_overlaps ?
        RGFA::Alignment::Placeholder.new : self.overlaps[i]
      retval << [self.segment_names[i], self.segment_names[j], cigar]
    end
    retval
  end

  # Are the overlaps a single "*"? This is a compact representation of
  # a linear path where all CIGARs are "*"
  # @return [Boolean]
  # @api private
  def undef_overlaps?
    overlaps.size == 1 and overlaps[0].empty?
  end

  def initialize_references
    initialize_links
    initialize_segments
  end

  def initialize_links
    refs[:links] = []
    compute_required_links.each do |from,to,cigar|
      l = nil
      orient = :+
      if @rgfa.segment(from.line) and @rgfa.segment(to.line)
        l = @rgfa.search_link(from, to, cigar)
        if !l.nil? and l.compatible_complement?(from, to, cigar)
          orient = :-
        end
      end
      if l.nil?
        if @rgfa.segments_first_order
          raise RGFA::NotFoundError, "Path: #{self}\n"+
          "requires a non-existing link:\n"+
          "from=#{from} to=#{to} cigar=#{cigar}"
        end
        l = RGFA::Line::Edge::Link.new({:from => from.line,
                                  :from_orient => from.orient,
                                  :to => to.line,
                                  :to_orient => to.orient,
                                  :overlap => cigar},
                                  virtual: true,
                                  version: :gfa1)
        l.connect(@rgfa)
      end
      @refs[:links] << OL[l,orient]
      l.add_reference(self, :paths)
    end
  end

  def initialize_segments
    segment_names.each do |sn_with_o|
      s = @rgfa.segment(sn_with_o.line)
      sn_with_o.line = s
      s.add_reference(self, :paths)
    end
  end

  def backreference_keys(ref, key_in_ref)
    case ref.record_type
    when :L
      [:links]
    when :S
      [:segment_names]
    end
  end

end
